src/queue/priority.rs

```diff
@@ -0,0 +1,501 @@
+    pub fn push(&self, mut task_cell: T) {
+        let priority = self.task_manager.prepare_before_push(&mut task_cell);
+        set_schedule_time(&mut task_cell);
+        self.queue.push(task_cell, priority);
+    }
+}
+
+/// The local queue is just a proxy of the global queue.
+pub(crate) struct LocalQueue<T> {
+    queue: Arc<QueueCore<T>>,
+    task_manager: PriorityTaskManager,
+}
+
+impl<T> LocalQueue<T>
+where
+    T: TaskCell + Send,
+{
+    pub(super) fn push(&mut self, mut task_cell: T) {
+        let priority = self.task_manager.prepare_before_push(&mut task_cell);
+        set_schedule_time(&mut task_cell);
+        self.queue.push(task_cell, priority);
+    }
+
+    pub(super) fn pop(&mut self) -> Option<Pop<T>> {
+        self.queue.pop()
+    }
+
+    pub(super) fn has_tasks_or_pull(&mut self) -> bool {
+        !self.queue.is_empty()
+    }
+}
+
+/// A trait used to generate priority value for each task.
+pub trait TaskPriorityProvider: Send + Sync + 'static {
+    /// Return a priority value of this task, all tasks in the priority queue is ordered by this value.
+    fn get_priority(&self, extras: &Extras) -> u64;
+}
+
+///
+#[derive(Clone)]
+struct PriorityTaskManager {
+    level_manager: Arc<TaskLevelManager>,
+    priority_manager: Arc<dyn TaskPriorityProvider>,
+}
+
+impl PriorityTaskManager {
+    fn prepare_before_push<T>(&self, task_cell: &mut T) -> u64
+    where
+        T: TaskCell,
+    {
+        self.level_manager.adjust_task_level(task_cell);
+        self.priority_manager.get_priority(task_cell.mut_extras())
+    }
+}
+
+/// The global priority queue. We use a `SkipMap` as a priority queue,
+/// The key is the priority and value is the task.
+struct QueueCore<T> {
+    pq: SkipMap<MapKey, Slot<T>>,
+    /// a global sequence generator to ensure all task keys are unique.
+    sequence: AtomicU64,
+}
+
+impl<T> QueueCore<T> {
+    fn new() -> Self {
+        Self {
+            pq: SkipMap::new(),
+            sequence: AtomicU64::new(0),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.pq.is_empty()
+    }
+}
+
+impl<T: TaskCell + Send + 'static> QueueCore<T> {
+    fn push(&self, msg: T, priority: u64) {
+        self.pq.insert(self.gen_key(priority), Slot::new(msg));
+    }
+
+    pub fn pop(&self) -> Option<Pop<T>> {
+        fn into_pop<T>(mut t: T) -> Pop<T>
+        where
+            T: TaskCell,
+        {
+            let schedule_time = t.mut_extras().schedule_time.unwrap();
+            Pop {
+                task_cell: t,
+                schedule_time,
+                from_local: false,
+            }
+        }
+
+        self.pq
+            .pop_front()
+            .map(|e| into_pop(e.value().take().unwrap()))
+    }
+
+    #[inline]
+    fn gen_key(&self, weight: u64) -> MapKey {
+        MapKey(weight, self.sequence.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+/// A holder to store task. The slot can be concurrently visit by multiple thread in the skip-list.
```

Comment:
I only see `push` and `pop` access the skiplist and they don't seem to visit a slot concurrently. Where does concurrent visit happen?

Reply:
Yes, in our case it's safe to take the value. But the pop() api only provide readonly access to the key and value. So here we wrap it in a atomic ptr to workaround this restriction. Do you have any idea to avoid this?

Reply:
Seems https://github.com/crossbeam-rs/crossbeam/blob/0de660ec88481ff7787374efee0c2e21d4d403b2/crossbeam-utils/src/atomic/atomic_cell.rs#L1020 AtomicCell is also not very check because: 1. It uses a global lock for large value (> 8 bytes), 2. We need to use `AtomicCell::swap` to take the value which new to swap in a default TaskCell but new a TaskCell(with dummy Future) also needs 1 box allocation.

Reply:
Ok. I found we can simple replace the AtomicPtr<T> with `RefCell<Option<T>>`. Since only the thread pop the entry can visit the value, it safe to impl Sync to the Slot. @sticnarf  PTAL again.

Reply:
The future taskcell is only 8 bytes so atomic cell is cheap. For other types, yes, a `Mutex` is better than the implementation of `AtomicCell`.

While I don't think `RefCell` + `unsafe impl Sync` is a good combination.

If you care about safety, `Mutex<Option<>>` is your choice. And it's fast in practice when there is not any race (only need a very few atomic operations).
If you are confident enough, just use `UnsafeCell`. `RefCell` does not provide you with extra safety when you mark it `Sync` because its borrow checks will become unreliable when multiple threads access it concurrently.

Reply:
Done. Change to using `AtomicCell`.